test/creds_test.go

```diff
@@ -545,3 +552,177 @@ func (s) TestServerCredsDispatch(t *testing.T) {
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		BasicConstraintsValid: true,
+	}
+
+	// if ca is nil, self-sign the certificate
+	var (
+		parentCert *x509.Certificate = template
+		parentKey  interface{}       = key
+	)
+	if ca != nil {
+		parentCert, err = x509.ParseCertificate(ca.Certificate[0])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse CA certificate: %v", err)
+		}
+		parentKey = ca.PrivateKey
+	}
+
+	certData, err := x509.CreateCertificate(rand.Reader, template, parentCert, &key.PublicKey, parentKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create certificate: %v", err)
+	}
+
+	keyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	})
+	certPem := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certData,
+	})
+
+	cert, err := tls.X509KeyPair(certPem, keyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse public/private certs: %v", err)
+	}
+
+	return &cert, nil
+}
+
+func (s) TestClientCredsHandshakeFailure(t *testing.T) {
+	// load server certificate
+	cert, err := tls.LoadX509KeyPair(testdata.Path("x509/server1_cert.pem"), testdata.Path("x509/server1_key.pem"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// load server CA certificate
+	ca, err := tls.LoadX509KeyPair(testdata.Path("x509/server_ca_cert.pem"), testdata.Path("x509/server_ca_key.pem"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// create server tls config
+	roots := x509.NewCertPool()
+	roots.AppendCertsFromPEM(pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: ca.Certificate[0],
+	}))
+
+	tlsCfg := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientCAs:    roots,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	}
+
+	// start server listener
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer lis.Close()
+
+	// start the test server using the credentials
+	s := grpc.NewServer(grpc.Creds(credentials.NewTLS(tlsCfg)))
+	go s.Serve(lis)
+	defer s.Stop()
+
+	// pre-generate client certificates
+	clientCerts := clientCertificates{}
+	if err := clientCerts.generateCertificates(&ca); err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		cert          *tls.Certificate
+		shouldFail    bool
+		expectedError string
+	}{
+		{&tls.Certificate{}, true, "remote error: tls: bad certificate"},
+		{clientCerts.expiredCert, true, "remote error: tls: bad certificate"},
+		{clientCerts.selfSignedCert, true, "remote error: tls: bad certificate"},
+		{clientCerts.validCert, false, ""},
+	}
+
+	for i, test := range tests {
+		cfg := &tls.Config{
+			Certificates:       []tls.Certificate{*test.cert},
+			InsecureSkipVerify: true, // not intrested in server certificates
+		}
+		creds := credentials.NewTLS(cfg)
+
+		dialCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		cc, err := grpc.DialContext(
```

Comment:
Can you please also add a check for RPC errors:

```go
cc, err := grpc.Dial(.../* no WithBlock or WithReturnConnectionError*/)
...
client := testpb.NewTestServiceClient(cc)
res, err := client.EmptyCall(...)
// Ensure err contains the connection error
```

You should be able to use the `stubserver.StubServer` to make it easy to create the client and server (server and dial options can be provided to the `Start` method); check the other tests for examples.